File: 1-two-sum/two-sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */

 //naive solutions O(n^2)
// var twoSum = (nums, target) => {
//     for(let i=0;i<nums.length;i++){
//         for(let j=1;j<nums.length;j++){
//             if(nums[i]+nums[j]===target){
//                 return [i,j]
//             }
//         }
//     }
// };


//optimised solution linear time O(n)
var twoSum = (nums, target) =>{
    const myObj ={};
    for(let i=0;i<nums.length;i++) {
        if(myObj.hasOwnProperty(nums[i])) {
            return [myObj[nums[i]],i];
        }
        myObj[target-nums[i]]=i;
    }

    return false;
}

File: 1-two-sum/two-sum.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int,int> hashMap;
        for(int i=0;i<nums.size();i++){
            if(hashMap.find(nums[i])!=hashMap.end()){
                return {i,hashMap[nums[i]]};
            }
            else {
                hashMap.insert({target-nums[i],i});
            }
        }
        return {};
    }
};